### PR TITLE
Assignment6 @lsobtafo

### PR DESCRIPTION
**Pull request by slanante**
Initial PR, will add more tests.
Had issues with sending a PR since the set up of the repository was incorrect.
Here are the commits from the original branch:
image

### CODE REVIEW

File: Calculate/Calculator.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+namespace Calculate
+{
+    public class Calculator
+    {
+        public static Func<int, int, int> Add { get; set; } = (v1, v2) => v1 + v2;
+        public static Func<int, int, int> Subtract { get; set; } = (v1, v2) => v1 - v2;
+        public static Func<int, int, int> Multiply { get; set; } = (v1, v2) => v1 * v2;
+        public static Func<int, int, int> Divide { get; set; } = (v1, v2) => v1 / v2;
```
Comment (jacobnewcomb) [inline Calculate/Calculator.cs:8]: I would maybe check if v2 is zero here. If so, throw an error?

**Review by jacobnewcomb — COMMENTED** [review]
- Create a Console project called "Calculate.". ✔
- Define a Program Class
        Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
        Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
        Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
        Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
        Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
            is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
                Uses char for the key corresponding to the operators +, -, *, and /. ✔
                Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
        Implement a TryCalculate method following "TryParse" pattern ✔
            Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
            If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
            Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
            Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

File: Calculate/Calculator.cs (hunk 1 above)
Comment (jacobnewcomb) [reply Calculate/Calculator.cs:8]: for the division specifically

Comment (slanante) [reply Calculate/Calculator.cs]: Cool, will do thank you!

**Review by mmwoodfo — COMMENTED** [review]
image
Why can't you handle fractions?

image
This input is valid, and would have had a similar output to the one above (decimal)

File: Calculate/Calculator.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        public static Func<int, int, int> Multiply { get; set; } = (v1, v2) => v1 * v2;
+        public static Func<int, int, int> Divide { get; set; } = (v1, v2) =>
+        {
+            if (v2 == 0)
+            {
+                throw new DivideByZeroException("Cannot divide by zero");
+            }
+            if (v1 < v2)
+            {
+                throw new InvalidOperationException("Numerator must be greater than or equal to the denominator, we cannot handle fractions");
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:16]: Why? This should be handled.

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                return false;
+            }
+
+            if (!Calculator.MathematicalOperations.TryGetValue(symbol, out var operation))
+            {
+                return false;
+            }
+
+            if (char.TryParse(parts[1], out symbol).Equals('/') && (parts[2] == "0"))
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:59]: Isn't this handled by your catch below?

File: Calculate/ProgramBase.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+namespace Calculate
+{
+    public abstract class ProgramBase
```
Comment (mmwoodfo) [inline Calculate/ProgramBase.cs:3]: This should have a constructor to set these properties, as well as a default constructor to allow for the default implementation.

File: Calculate/Calculator.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
+namespace Calculate
+{
+    public class Calculator
+    {
+        public static Func<int, int, int> Add { get; set; } = (v1, v2) => v1 + v2;
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:5]:
```suggestion
        public static readonly Func<int, int, int> Add { get; set; } = (v1, v2) => v1 + v2;
```

File: Calculate.Tests/Calculate.Tests.csproj [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (mmwoodfo) [inline Calculate.Tests/Calculate.Tests.csproj:4]: Should be taken care of in your directory build props file

File: Calculate/Calculator.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                {'-', Subtract},
+                {'*', Multiply},
+                {'/', Divide}
+            };
+
+        public static bool TryCalculate(string input, out int result)
+        {
+            result = 0;
+
+            string[] parts = input.Split(' ');
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:35]: You should be using the Split method along side pattern matching. e.g.
`if (expression?.Split(" ") is [string operator1, [char mathmaticalOperator], string operand2])`

**Pull request closed without merging**